### PR TITLE
point `path fetch` to chain-reg and avoid overwriting channel-filters

### LINKED ISSUE #680: running `rly paths fetch` overwrites any path filtering options user has in-place
When running `fetch`
We should check to see if path already exists in config, if so, it's not necessary to "re-add" it.

### PR DESCRIPTION
**Pull request by boojamya**
This PR updates the `rly paths fetch` command and how it writes to the config.

- Fetches path meta data from cosmos/chain-registry/_IBC as opposed to cosmos/relayer/interchain folder

- Since paths are bi-directional, it only adds the paths once (Is this ok with interchain accounts?)

- instead of abbreviated paths like hubosmo, paths are now added based off the chain-registry file name ie. cosmoshub-osmosis ( with checks implemented on the chain-registry side, these will always be in alphabetical order)

- the fetch command no longer overwrites channel filters in the config unless flag "-o" is added. Closes https://github.com/cosmos/relayer/issues/680

### CODE REVIEW

File: cmd/paths.go [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 	"io/ioutil"
-	"os"
-	"path"
 	"path/filepath"
+	"sort"
 	"strings"
 
 	"github.com/cosmos/relayer/v2/relayer"
-	"github.com/go-git/go-git/v5"
+	"github.com/google/go-github/github"
```
Comment (mark-rushakoff) [inline cmd/paths.go]: This should be `github.com/google/go-github/v43/github` to match the version of the module we are using elsewhere in the project.

As a side note, it looks like the current release is v45, but updating to that is outside the scope of this PR.

File: cmd/paths.go [hunk 2]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
-			}); err != nil {
-				return err
-			}
+			combinations := permutations(chains)
 
-			// Try to fetch path info for each configured chain that has canonical chain/path info in the GH repo
-			for _, srcChain := range a.Config.Chains {
-				for _, dstChain := range a.Config.Chains {
+			client := github.NewClient(nil)
+			gh_ctx := context.Background()
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Instead of `context.Background()`, this should use the context associated with `cmd`, at `cmd.Context()`. Its use can be inlined in the one call, no need for this variable.

File: cmd/paths.go [hunk 3]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
-							if err = a.Config.AddPath(pthName, p); err != nil {
-								return fmt.Errorf("failed to add path %s: %w", pth, err)
-							}
-
-							fmt.Fprintf(cmd.ErrOrStderr(), "added path %s...\n", pthName)
-						}
-					}
+				_, exist := a.Config.Paths[pthName]
+				if exist && !overwrite {
+					fmt.Printf("skipping:  %s already exists in config, use -o to overwrite (clears filters) \n", pthName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: `a.Log.Info` or `fmt.Fprintf(cmd.ErrOrStderr(), "...")` here.

File: cmd/paths.go [hunk 4]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					continue
+				}
 
-					if err := a.OverwriteConfig(a.Config); err != nil {
-						return err
+				// TDOD: Don't use github api. Potentially use: https://github.com/eco-stake/cosmos-directory once they integrate IBC data into restAPI. This will avoid rate limits.
```
Comment (mark-rushakoff) [inline cmd/paths.go]: https://www.abbreviations.com/TDOD 😜

File: cmd/paths.go [hunk 5]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+				fileName := pthName + ".json"
+				reg_path := filepath.Join("_IBC", fileName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: This should be `path.Join`; `filepath` is OS-specific and therefore will not work properly on Windows.

Don't use underscores in variable names in Go.

File: cmd/paths.go [hunk 6]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+				r, err := client.Repositories.DownloadContents(gh_ctx, "cosmos", "chain-registry", reg_path, nil)
+				if err != nil {
+					if _, ok := err.(*github.RateLimitError); ok {
```
Comment (mark-rushakoff) [inline cmd/paths.go]: The relatively new convention for this is `errors.As(err, new(github.RateLimitError))`, as that handles wrapped errors properly. OTOH, I suppose we should be able to rely on the github methods returning concrete instances.

File: cmd/paths.go [hunk 7]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+						return fmt.Errorf("hit github rate limit ERR: %w", err)
 					}
+					fmt.Printf("not found:  path %s not found in repo 'cosmos/chain-registry', folder '_IBC'\n", pthName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Use `a.Log` or `fmt.Fprintf(cmd.ErrOrStderr()` here; and include the received error instead of assuming the error is not found.

File: cmd/paths.go [hunk 8]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					continue
+				}
+
+				bytes, err := ioutil.ReadAll(r)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: `b, err := io.ReadAll(r)` -- `bytes` overlaps with the `bytes` package, and the `ioutil` package is deprecated.

File: cmd/paths.go [hunk 9]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					fmt.Printf("skipping:  %s already exists in config, use -o to overwrite (clears filters) \n", pthName)
+					continue
+				}
 
-					if err := a.OverwriteConfig(a.Config); err != nil {
-						return err
+				// TDOD: Don't use github api. Potentially use: https://github.com/eco-stake/cosmos-directory once they integrate IBC data into restAPI. This will avoid rate limits.
+				fileName := pthName + ".json"
+				reg_path := filepath.Join("_IBC", fileName)
+				r, err := client.Repositories.DownloadContents(gh_ctx, "cosmos", "chain-registry", reg_path, nil)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: The docs for DownloadContents mention:

> It is the caller's responsibility to close the ReadCloser.

Since there may be some early returns, it is better to have `defer r.Close()` immediately after the error check, and then you may still have an explicit `r.Close()` after reading it. In this case, we can safely ignore the error returned from `r.Close()`.

File: cmd/paths.go [hunk 10]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					if _, ok := err.(*github.RateLimitError); ok {
+						return fmt.Errorf("hit github rate limit ERR: %w", err)
 					}
+					fmt.Printf("not found:  path %s not found in repo 'cosmos/chain-registry', folder '_IBC'\n", pthName)
+					continue
+				}
+
+				bytes, err := ioutil.ReadAll(r)
+				if err != nil {
+					fmt.Printf("error reading response body: %v", err)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Returning the error is appropriate (otherwise we will continue with missing or garbage data).

File: cmd/paths.go [hunk 11]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+				}
+
+				ibc := &relayer.IBC_data{}
+				if err = json.Unmarshal(bytes, &ibc); err != nil {
+					fmt.Println("failed to unmarshal ", err)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Return the error.

File: cmd/paths.go [hunk 12]
```diff
@@ -256,77 +256,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+				newPath := &relayer.Path{
+					Src: srcPathEnd,
+					Dst: dstPathEnd,
+				}
+				r.Close()
+
+				if err = a.Config.AddPath(pthName, newPath); err != nil {
+					return fmt.Errorf("failed to add path %s: %w", pthName, err)
+				}
+				fmt.Printf("added:  %s\n", pthName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Log or Fprintf.

File: relayer/path.go [hunk 13]
```diff
@@ -101,6 +101,38 @@ type ChannelFilter struct {
 	ChannelList []string `yaml:"channel-list" json:"channel-list"`
 }
 
+type IBC_data struct {
```
Comment (mark-rushakoff) [inline relayer/path.go]: No underscore in type name.

**Review by mark-rushakoff — COMMENTED** [review]
There are a few things that need cleaned up and then we can take another look.

File: cmd/paths.go (hunk 3 above)
Comment (boojamya) [reply cmd/paths.go]: Went with Fprintf here since I don't think zapper is necessary for this output.

File: cmd/paths.go (hunk 9 above)
Comment (boojamya) [reply cmd/paths.go]: After initial attempt, I'm having trouble with this one. 
Having a `defer r.Close()` immediately after error check is panicking. 
`panic: runtime error: invalid memory address or nil pointer dereference`

I may be doing something wrong here.

Comment (jtieri) [reply cmd/paths.go]: > After initial attempt, I'm having trouble with this one. Having a `defer r.Close()` immediately after error check is panicking. `panic: runtime error: invalid memory address or nil pointer dereference`
> 
> I may be doing something wrong here.

You'll want to move the call to `defer` to after the check on the error

File: cmd/paths.go [hunk 14]
```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
-							if err = a.Config.AddPath(pthName, p); err != nil {
-								return fmt.Errorf("failed to add path %s: %w", pth, err)
-							}
-
-							fmt.Fprintf(cmd.ErrOrStderr(), "added path %s...\n", pthName)
-						}
-					}
+				_, exist := a.Config.Paths[pthName]
+				if exist && !overwrite {
+					// fmt.Printf("skipping:  %s already exists in config, use -o to overwrite (clears filters) \n", pthName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Comment can be removed.

File: cmd/paths.go [hunk 15]
```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
-						return err
+				// TODO: Don't use github api. Potentially use: https://github.com/eco-stake/cosmos-directory once they integrate IBC data into restAPI. This will avoid rate limits.
+				fileName := pthName + ".json"
+				regPath := path.Join("_IBC", fileName)
+				reader, _, err := client.Repositories.DownloadContents(cmd.Context(), "cosmos", "chain-registry", regPath, nil)
+				if err != nil {
+					if errors.As(err, new(*github.RateLimitError)) {
+						return fmt.Errorf("github API rate limit reached while querying data ERR: %w", err)
 					}
+					fmt.Fprintf(cmd.ErrOrStderr(), "not found:  path %s not found in repo 'cosmos/chain-registry', folder '_IBC'\n", pthName)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: There are likely other errors that may occur here besides "not found" -- maybe use something more like `"failure retrieving %s: %v", pthName, err`.

File: cmd/paths.go [hunk 16]
```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					continue
+				}
+
```
Comment (mark-rushakoff) [inline cmd/paths.go:299]: As Justin mentioned, after handling the `err != nil` case, then you can put `defer reader.Close()`.

File: cmd/paths.go [hunk 17]
```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
 			}
 
-			if _, err = git.PlainClone(localRepo, false, &git.CloneOptions{
-				URL:           REPOURL,
-				Progress:      io.Discard,
-				ReferenceName: "refs/heads/main",
-			}); err != nil {
-				return err
-			}
+			combinations := permutations(chains)
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Pedantically, combinations and permutations are not the same thing.

File: cmd/paths.go [hunk 18]
```diff
@@ -256,77 +255,124 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
 			return nil
+
 		},
 	}
-	return cmd
+	return OverwriteConfigFlag(a.Viper, cmd)
+}
+
+// used in pathsFetch command to find all possible combinations of paths for configured chains
+func permutations(arr []string) []string {
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Would you mind using https://pkg.go.dev/github.com/cespare/permute/v2 instead?

**Review by mark-rushakoff — COMMENTED** [review]
A couple more small things to clean up, and then I think this can ship.

File: cmd/paths.go (hunk 18 above)
Comment (boojamya) [reply cmd/paths.go]: This package seemed great, however I am unable to get the permutations outputed by Permute() into a slice of slices.

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		x := chain_list
		fmt.Println(x)
		permutations = append(permutations, x)
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```
```
[Cat Dog Snake]
[Dog Cat Snake]
[Snake Cat Dog]
[Cat Snake Dog]
[Dog Snake Cat]
[Snake Dog Cat]
PERMUTATIONS HERE:  [[Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat] [Snake Dog Cat]]
```
The printed slice is different than what gets appended to the array. 
I feel like I'm missing something simple here.

File: cmd/paths.go (hunk 17 above)
Comment (DavidNix) [reply cmd/paths.go]: https://medium.com/i-math/combinations-permutations-fa7ac680f0ac

> With permutations we care about the order of the elements, whereas with combinations we don’t.

File: cmd/paths.go (hunk 18 above)
Comment (DavidNix) [reply cmd/paths.go]: Go slices are tricky because they reference the same backing array even when you assign a new variable like `x`. You can think of slices as a "view" of an array. 

https://go.dev/blog/slices-intro
> re-slicing a slice doesn’t make a copy of the underlying array

In this case `chain_list` and `x` are sharing the same backing array which is why your `PERMUTATIONS HERE:` prints only the last combination of `Snake Dog Cat`. 

(Side note: I believe these are actually combinations, not permutations because order doesn't matter.)

The solution is you must explicitly allocate a new array. 

A couple options that work:

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		fmt.Println(chain_list)
		x := make([]string, len(chain_list))
		copy(x, chain_list)
		permutations = append(permutations, x)
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```

Or less lines of codes:

```go
	chain_list := []string{"Cat", "Dog", "Snake"}
	p := permute.Slice(chain_list)
	var permutations [][]string
	for p.Permute() {
		fmt.Println(chain_list)
		permutations = append(permutations, append([]string{}, chain_list...))
	}
	fmt.Println("PERMUTATIONS HERE: ", permutations)
```

Go's reason for this, AFAIK, is to conserve memory. Go won't allocate a new array unless it has to.

Fyi, I use this a lot: https://ueokande.github.io/go-slice-tricks/

Comment (boojamya) [reply cmd/paths.go]: Thanks for the explanation here @DavidNix!
Got this working.

File: cmd/paths.go [hunk 19]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
-
-							fmt.Fprintf(cmd.ErrOrStderr(), "added path %s...\n", pthName)
-						}
+				// TODO: Don't use github api. Potentially use: https://github.com/eco-stake/cosmos-directory once they integrate IBC data into restAPI. This will avoid rate limits.
+				fileName := pthName + ".json"
+				regPath := path.Join("_IBC", fileName)
+				client, _, err := client.Repositories.DownloadContents(cmd.Context(), "cosmos", "chain-registry", regPath, nil)
+				if err != nil {
+					if errors.As(err, new(*github.RateLimitError)) {
+						return fmt.Errorf("github API rate limit reached while querying data ERR: %w", err)
```
Comment (jtieri) [inline cmd/paths.go]: You could drop the `ERR:` part at the end of this error message so it's just `error message: %w`, this is the convention we have started to use.

File: cmd/paths.go [hunk 20]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
 $ %s paths fetch --home %s
 $ %s pth fch`, appName, defaultHome, appName)),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			// Clone the GH repo to tmp dir, we will extract the path files from here
-			localRepo, err := os.MkdirTemp("", "")
-			if err != nil {
-				return err
+			chains := []string{}
+			for chain_name := range a.Config.Chains {
```
Comment (jtieri) [inline cmd/paths.go]: I would opt for `camelCase` variable names instead of `snake_case`

File: cmd/paths.go [hunk 21]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+					continue
+				}
+				defer client.Close()
 
-					if err := a.OverwriteConfig(a.Config); err != nil {
-						return err
-					}
+				b, err := io.ReadAll(client)
+				if err != nil {
+					return fmt.Errorf("error reading response body: %v", err)
```
Comment (jtieri) [inline cmd/paths.go]: I would use `%w` here instead of `%v`

This wraps the error so that it can be unwrapped with calls to `errors.Unwrap` and tested with calls to `errors.Is` and `errors.As`

File: cmd/paths.go [hunk 22]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+				}
+
+				ibc := &relayer.IBCdata{}
+				if err = json.Unmarshal(b, &ibc); err != nil {
+					return fmt.Errorf("failed to unmarshal: %v ", err)
```
Comment (jtieri) [inline cmd/paths.go]: I would use `%w` here instead of `%v` as well

**Review by jtieri — COMMENTED** [review]
A few small nits but otherwise this looks good to go now

File: cmd/paths.go [hunk 23]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
+			var chainCombinations []string
+			for p.Permute() {
+				tmp := make([]string, 2)
+				copy(tmp, chains)
+				sort.Strings(tmp)
+				tmp1 := string(tmp[0] + "-" + tmp[1])
+				if !contains(chainCombinations, tmp1) {
+					chainCombinations = append(chainCombinations, tmp1)
+				}
 			}
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Here's how I would write this loop:

```go
// Store combinations in a map so we don't have to repeatedly
// scan for entry existence.
chainCombinations := make(map[string]bool)

for p.Permute() {
  a, b := chains[0], chains[1]

  // No need to use a sorting algorithm for a pair of items.
  // The Go compiler should be smart enough to only assign the string once.
  pair := a + "-" + b
  if b < a {
    pair = b + "-" + a
  }

  chainComboMap[pair] = true
}

// ...

for pathName := range chainComboMap {
  // ...
}
```

File: go.mod [hunk 24]
```diff
@@ -19,6 +18,7 @@ require (
 	github.com/avast/retry-go/v4 v4.0.5
 	github.com/cosmos/ibc-go/v3 v3.0.0
 	github.com/google/go-cmp v0.5.8
+	github.com/google/go-github v17.0.0+incompatible
```
Comment (mark-rushakoff) [inline go.mod]: Not sure why v17 is back in go.mod -- maybe one more `go mod tidy`?

File: cmd/paths.go [hunk 25]
```diff
@@ -2,16 +2,17 @@ package cmd
...
 	"path"
-	"path/filepath"
+	"sort"
 	"strings"
 
+	"github.com/cespare/permute/v2"
 	"github.com/cosmos/relayer/v2/relayer"
-	"github.com/go-git/go-git/v5"
+	"github.com/google/go-github/v43/github"
+
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Remove blank line so non-standard-library imports are grouped together.

File: cmd/paths.go [hunk 26]
```diff
@@ -256,77 +257,100 @@ func pathsFetchCmd(a *appState) *cobra.Command {
...
 			}
 
-			if _, err = git.PlainClone(localRepo, false, &git.CloneOptions{
-				URL:           REPOURL,
-				Progress:      io.Discard,
-				ReferenceName: "refs/heads/main",
-			}); err != nil {
-				return err
+			//find all combinations of paths for configured chains
+			//note: path file names on the chain-registry are in alphabetical order ex: achain-zchain.json
```
Comment (mark-rushakoff) [inline cmd/paths.go]: Go comments have a space after `//`.

**Review by mark-rushakoff — APPROVED** [review]
There are a couple small style cleanups I would do, but overall I think this is good to go.

File: cmd/paths.go (hunk 23 above)
Comment (boojamya) [reply cmd/paths.go]: Thats beautiful
